### PR TITLE
fix(core, processor): Resolve MAST hash collisions (close #605)

### LINKED ISSUE #605: Proof of 0 == 1
When you execute the following Miden Assembly program, you end up with a stack of all 0s:

```
begin
  if.true
    push.1
  end
end
```

However this [repository](https://github.com/matthiasgoergens/miden-collision) has a proof to show that the execution ends with a 1 on the stack.

### PR DESCRIPTION
**Pull request by ekmett**
## Describe your changes

This patch applies a tweak to the hashing function used for MAST nodes to ensure different node types hash to different values. This ensures that the simple exploit from #605 is no longer sufficient to subvert the verification process.

The hashes themselves are transformed by a simple transformation by postmultiplying by a constant chosen based on the node type. This prevents different node types from hashing to the same value. These coefficients were arbitrarily chosen to be sequential primes. 2 was avoided in case the base hash ever switches from elements to something in base "user" numerical range 2^32. Rescue-Prime should by the strong hashing assumption have no meaningful interactions with multiplication by a non-0 constant.

Further, syscalls and calls now hash to different values.

Documentation has been updated to reflect the change.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
Please merge or rebase on latest `next` to make the clippy failure go away.

**Comment by ekmett** [comment]
@matthiasgoergens beat me to it!

File: core/src/program/blocks/call_block.rs [hunk 1]
```diff
@@ -36,7 +40,7 @@ impl Call {
     /// as a kernel call.
     pub fn new_syscall(fn_hash: Digest) -> Self {
         // TODO: make hash computation different from regular call
```
Comment (matthiasgoergens) [inline core/src/program/blocks/call_block.rs:42]: You probably want to remove this TODO.

**Comment by bobbinth** [comment]
Closing this as this has been superseded by #682. Would be great to confirm that #605 has been fixed though.

**Pull request closed without merging**